Introduce `<Notice>` component for WooCommerce Admin

corsonr opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR is part of the Sale banner project p9h1Lv-bHt-p2  

Closes https://github.com/Automattic/woocommerce.com/issues/19573 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch
2. Open `plugins/woocommerce-admin/client/marketplace/components/content/content.tsx` and import the `<Notice>` component with `import Notice from '../notice/notice';`
3. At line **141** add the component with
```jsx
			<Notice
				id="notice-test"
				variant="info"
				description="This is a <strong>test</strong>."
				icon="info"
				isDismisible
			/>
			<Notice
				id="notice-test-2"
				variant="warning"
				description="This is a test."
				icon="info"
				isDismisible
			/>
			<Notice
				id="notice-test-3"
				variant="success"
				description="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse urna magna, cursus eget orci convallis, ullamcorper tincidunt urna. Pellentesque blandit hendrerit risus, in mattis tellus. Nulla faucibus metus rutrum, commodo nunc non, pretium eros. Curabitur libero sapien, accumsan nec ullamcorper eu, rutrum vel odio. Aliquam in nisi augue. Vestibulum sit amet auctor mi. Praesent ligula nisi, dignissim ac nisi vitae, ullamcorper porttitor lacus. Mauris quis lacinia lacus. Sed id magna sed odio rhoncus sagittis ut vitae leo. Aliquam laoreet mollis adipiscing. Vivamus viverra, dui eget fringilla consequat, tellus felis ornare orci, sit amet hendrerit diam nibh et diam. Quisque metus arcu, tristique at ultricies id, venenatis sit amet risus. Fusce et tempus nulla. Aliquam vitae sem quis enim varius auctor lacinia at mi. Nulla facilisi."
				icon="check"
				isDismisible={ false }
			/>
```
4. run `pnpm build`
5. visit the [in-app marketplace](http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions) page
6. confirm you see the notices
<img width="1072" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1649788/37143f65-ca9e-416c-8d05-b279df2635de">

7. confirm mobile view is working as expected
8. Close one of the two dismissible notices, and confirm in the `Application > Local Storage` section of your browser console that an entry is saved for that notice. Refresh the page and confirm the hidden notice doesn't show anymore.
9. test various combinations of props as per the README file

<img width="1084" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1649788/87357e5e-586d-4e31-9d73-8301f409e063">


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

corsonr committed:
Introduce Notice component

corsonr committed:
minor alignment fix

corsonr committed:
Fix MD title increment

corsonr committed:
Fix readme lint

corsonr committed:
add changelog entry

corsonr committed:
fix CSS lints

andfinally commented:
Looking good! I'd like to propose a few minor tweaks.

* Regularising the spelling of `isDismissible`.
* Using `sanitizeHTML` to sanitize the notice description.
* Making the variant classes like `&-success` to make them a bit shorter.

[tweaks.patch](https://github.com/woocommerce/woocommerce/files/14334227/tweaks.patch)

andfinally committed:
Small tweaks.

Regularising the spelling of isDismissible.
Using sanitizeHTML to sanitize the notice description.
Making the variant classes like &-success to make them a bit shorter.
Adjusting wording of README to convey that this component is designed for the marketplace, though it can be used elsewhere.
Tweaking whitespace in one place in the TSX to please the linter.

mcliwanow commented on plugins/woocommerce-admin/client/marketplace/components/notice/README.md:
```diff
@@ -0,0 +1,154 @@
+			} }
+			text="Label"
+		/>
+	</>
```
Fragment seems to be redundant

```suggestionx
	<Button
		variant="secondary"
		onClick={ () => {
			console.log( 'Primary button clicked' );
		} }
		text="Label"
	/>
	<Button
		variant="tertiary"
		onClick={ () => {
			console.log( 'Secondary button clicked' );
		} }
		text="Label"
	/>x
```

mcliwanow commented on plugins/woocommerce-admin/client/marketplace/components/notice/README.md:
```diff
@@ -0,0 +1,154 @@
+	icon="check"
+	isDismissible
+>
+	<>
```
Here too we can do without the fragment

andfinally reviewed (commented):
Looks great, thanks! I just had a few styling suggestions.

The margin between notices should be 32px on mobile.

I'd add a right margin of 12px on `.components-button` within a notice to keep buttons clear of each other.

I'd suggest we remove the 3px top margin from `.woocommerce-marketplace__notice-icon` and add `align-self: flex-start` to `.woocommerce-marketplace__notice-close` to make sure these elements align with the top of the text when the content is more than one line.

<img width="817" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/89f5e9fb-22fa-4299-9897-caefe2c996ae">

The close button looks especially odd centred when there is a lot of content.

<img width="320" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1647564/c5eab438-a67b-4754-90fa-b10c7f049699">

mcliwanow commented on plugins/woocommerce-admin/client/marketplace/components/notice/notice.tsx:
```diff
@@ -0,0 +1,99 @@
+
+export interface NoticeProps {
+	id: string;
+	children?: JSX.Element;
```
Changing the type to `React.ReactNode` would allow passing multiple children without wrapping them in a fragment:

```suggestion
	children?: React.ReactNode;
```

mcliwanow commented on plugins/woocommerce-admin/client/marketplace/components/notice/notice.tsx:
```diff
@@ -0,0 +1,99 @@
+		localStorage.setItem( `wc-marketplaceNoticeClosed-${ id }`, 'true' );
+	};
+
+	if ( ! isVisible ) return <></>;
```
let's return `null` instead. It's pretty common to return `null` from the components if they are supposed to render nothing

mcliwanow commented on plugins/woocommerce-admin/client/marketplace/components/notice/notice.tsx:
```diff
@@ -0,0 +1,99 @@
+	if ( ! isVisible ) return <></>;
+
+	const classes = classNames( 'woocommerce-marketplace__notice', {
+		'woocommerce-marketplace__notice-warning': variant === 'warning',
+		'woocommerce-marketplace__notice-info':
+			variant === 'info' || variant === 'info' || variant === 'error',
+		'woocommerce-marketplace__notice-error': variant === 'error',
+		'woocommerce-marketplace__notice-success': variant === 'success',
```
I wonder if all those optional clas names could be replaced with something like: `woocommerce-marketplace__notice-${ variant }`?

mcliwanow commented on plugins/woocommerce-admin/client/marketplace/components/notice/notice.tsx:
```diff
@@ -0,0 +1,99 @@
+
+	// Add a class for the icon color based on the variant for consistency.
+	const iconClass = classNames( 'woocommerce-marketplace__notice-icon', {
+		'icon-color-info': variant === 'info',
+		'icon-color-error': variant === 'error',
+		'icon-color-warning': variant === 'warning',
+		'icon-color-success': variant === 'success',
```
For consistency with the parent component class name I would suggest using the following string for a class name `woocommerce-marketplace__notice-icon--${ variant }`. Also, both here and in line 52, parts related to variant (info, error, warning, success) look like element [modifiers](https://getbem.com/naming/#modifier) and I see that it's a convention used in WooCommerce Core codebase, so maybe we should use it here too?

andfinally replied:
It also occurred to me we could do these as modifiers.

mcliwanow reviewed (commented):
The component itself looks great 🎉  I left a couple of comments that could improve the code though - things like changing prop types and using CSS class names more consistent with the rest of the codebase.

corsonr committed:
Update plugins/woocommerce-admin/client/marketplace/components/notice/notice.tsx

Co-authored-by: Michal Iwanow <4765119+mcliwanow@users.noreply.github.com>

corsonr committed:
address feedback

andfinally committed:
Changed notice classnames to use `__{$variant}` pattern as well. Restored 40px bottom margin to notices on viewports 600px and above. Added height 24px on icons to ensure they're nicely vertically centred.

andfinally reviewed (approved):
Looks 👌, thanks! I took the liberty of pushing 7420e78d004daab965e612fe71b4ac6ae29b0c32 with some small changes, hope that's OK!

- Using the `--{$variant}` pattern for the notices classnames also.
- Added `height: 24px;` on the icons to ensure they're nicely vertically centred.
- Restored 40px bottom margin to notices on viewports 600px and above.
